### PR TITLE
Stream atomic checkpoint saves to local disk instead of buffering

### LINKED ISSUE #21823: Avoid double-buffering checkpoints in memory when saving to local disk
### Outline & Motivation

_atomic_save in src/lightning/fabric/utilities/cloud_io.py currently serializes every checkpoint into an in-memory `io.BytesIO` buffer via `torch.save`, then writes that buffer's bytes out to the destination filesystem. For local-disk saves, this means the full checkpoint exists in memory twice at once: once as the live `state_dict` being serialized, and once as the serialized byte buffer being written to disk.

### Pitch

Change _atomic_save to branch on destination type:
  - Object storage (S3/GCS/Azure): keep the existing behavior.
  - Local/other filesystems: stream `torch.save(checkpoint, f)` directly into the open file handle, eliminating the redundant full in-memory buffer copy.

### Additional context

_No response_

cc @lantiga @justusschock

### PR DESCRIPTION
**Pull request by zhixiangli**
## What does this PR do?

Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/21823

Avoids holding a second full in-memory copy of the checkpoint when saving to local filesystems, which matters for large FSDP/ModelParallel full state dicts.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by deependujha — APPROVED** [review]

File: src/lightning/fabric/utilities/cloud_io.py [hunk 1]
```diff
@@ -95,28 +95,36 @@ def _atomic_save(checkpoint: dict[str, Any], filepath: _PATH) -> None:
...
+                else:
+                    # Use fs.pipe() for S3/GCS where it triggers parallel multipart uploads,
+                    # giving 4-5x throughput improvement for checkpoints >= 500 MB.
+                    fs.pipe(urlpath, bytesbuffer.getvalue())
             else:
+                # Stream directly to the file so we never hold a second full copy of the checkpoint
+                # in memory. This matters for large FSDP/ModelParallel full state dicts on local disk.
                 with fs.open(urlpath, "wb") as f:
-                    f.write(bytesbuffer.getvalue())
+                    torch.save(checkpoint, f)
```
Comment (deependujha) [inline src/lightning/fabric/utilities/cloud_io.py:127]: Hi @zhixiangli, just confirming: is `torch.save` atomic, if writing to fsspec file in `fs.transaction` context?

https://discuss.pytorch.org/t/is-torch-save-an-indivisible-transaction/141744

Comment (zhixiangli) [reply src/lightning/fabric/utilities/cloud_io.py:127]: fsspec/implementations/local.py: LocalFileOpener opens tempfile.mkstemp(). In the write path, writes go to that temporary file; commit() executes shutil.move(self.temp, self.path), and discard() deletes the temporary file.

Added the test test_atomic_save_local_interrupted_save_creates_no_partial_file.

**Review by tchaton — APPROVED** [review]
